plugin/pkg/auth/authorizer/node/node_authorizer.go

```diff
@@ -314,7 +314,7 @@ func (r *NodeAuthorizer) authorizeResourceSlice(nodeName string, attrs authorize
 	switch verb {
 	case "get", "create", "update", "patch", "delete":
 		// Okay, but check individual object permission below.
-	case "watch", "list":
+	case "watch", "list", "deletecollection":
```

Comment:
Is that a blocker and should I use a combination of "list" + "delete" instead?

Reply:
adding a guard to NodeRestriction admission plugin when deleting a ResourceSlice will protect things for now

adding a TODO here to restrict the deletecollection call to `fieldSelector=nodeName=$name` once that KEP is available will make sure the deletecollection call is made scoped

Reply:
> adding a guard to NodeRestriction admission plugin when deleting a ResourceSlice to make sure the node only deletes resource slices belonging to it will protect things for now

Sorry, I don't follow. Is that something that I need to add in this PR? Is the NodeRestriction admission plugin the node authorizer code that I'm extending in 2e34e187c994?

Reply:
> adding a guard to NodeRestriction admission plugin when deleting a ResourceSlice to make sure the node only deletes resource slices belonging to it will protect things for now (deletecollection handling invokes admission for each individual delete operation)

It does, but without passing in the object that is about to be deleted. I had implemented the admission check like this:

```patch
diff --git a/plugin/pkg/admission/noderestriction/admission.go b/plugin/pkg/admission/noderestriction/admission.go
index 9388ac52252..28cc2983205 100644
--- a/plugin/pkg/admission/noderestriction/admission.go
+++ b/plugin/pkg/admission/noderestriction/admission.go
@@ -641,15 +641,21 @@ func (p *Plugin) admitCSINode(nodeName string, a admission.Attributes) error {
 
 func (p *Plugin) admitResourceSlice(nodeName string, a admission.Attributes) error {
        // The create request must come from a node with the same name as the NodeName field.
-       // Other requests gets checked by the node authorizer.
-       if a.GetOperation() == admission.Create {
+       // Same when deleting an object.
+       //
+       // Other requests get checked by the node authorizer. The checks here are necessary
+       // because the node authorizer does not know the object content for a create request
+       // and not each deleted object in a DeleteCollection. DeleteCollection checks each
+       // individual object.
+       switch a.GetOperation() {
+       case admission.Create, admission.Delete:
                slice, ok := a.GetObject().(*resource.ResourceSlice)
                if !ok {
                        return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
                }
 
                if slice.NodeName != nodeName {
-                       return admission.NewForbidden(a, errors.New("can only create ResourceSlice with the same NodeName as the requesting node"))
+                       return admission.NewForbidden(a, errors.New("can only create or delete ResourceSlice with the same NodeName as the requesting node"))
                }
        }
 
```

But then tests failed. Adding debug output showed that this code gets called with a DELETE operation where `a.GetObject()` returns nil.

That means this stop-gap solution for securing deletecollection won't work and https://github.com/kubernetes/enhancements/pull/4600 is needed, right?

I'm adding a TODO and a tracking issue to move forward with this.

Reply:
The alternative is that kubelet does individual deletes. But that feels like a workaround. We can come back to it for DRA beta if https://github.com/kubernetes/enhancements/pull/4600 doesn't progress?

Reply:
> a.GetOldObject() works

:)

Reply:
Please also update node rules for rbac when the dra feature is enabled.

https://github.com/pohly/kubernetes/blob/bfe3552d5dc1675e33d535218437a5511d718f20/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L182

If the node authorizer is disabled by users, the dra feature in Kubelet will not work due to the missing permission.

cc @pohly

Reply:
Good point. "Node authorizer disabled" is not a configuration that is being tested right now.

Added. RBAC permissions for "resourceslices" were completely missing before, so we already had this problem in 1.30. The authorizer code allows other verbs because it's easier that way and a bit more flexible. For RBAC, I instead made it very specific to the exact verb that the kubelet is actually using. This can be changed more easily - unless I forget again about it.